mars/dataframe/groupby/sort.py

```diff
@@ -0,0 +1,184 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+import pandas as pd
+
+from ..operands import DataFrameOperandMixin
+from ..sort.psrs import DataFramePSRSChunkOperand
+from ... import opcodes as OperandDef
+from ...core import OutputType
+from ...core.operand import MapReduceOperand, OperandStage
+from ...serialization.serializables import (
+    Int32Field,
+    ListField,
+)
+from ...utils import (
+    lazy_import,
+)
+
+cudf = lazy_import("cudf", globals=globals())
+
+
+def _series_to_df(in_series, xdf):
+    in_df = in_series.to_frame()
+    if in_series.name is not None:
+        in_df.columns = xdf.Index([in_series.name])
+    return in_df
+
+
+class DataFrameGroupbyConcatPivot(DataFramePSRSChunkOperand, DataFrameOperandMixin):
+    _op_type_ = OperandDef.GROUPBY_SORT_PIVOT
+
+    @property
+    def output_limit(self):
+        return 1
+
+    @classmethod
+    def execute(cls, ctx, op):
+        inputs = [ctx[c.key] for c in op.inputs if len(ctx[c.key]) > 0]
+
+        xdf = pd if isinstance(inputs[0], (pd.DataFrame, pd.Series)) else cudf
+
+        a = xdf.concat(inputs, axis=0)
+        a = a.sort_index()
+        index = a.index.drop_duplicates()
+
+        p = len(inputs)
+        if len(index) < p:
+            num = p // len(index) + 1
+            index = index.append([index] * (num - 1))
+
+        index = index.sort_values()
+
+        values = index.values
+
+        slc = np.linspace(
+            p - 1, len(index) - 1, num=len(op.inputs) - 1, endpoint=False
+        ).astype(int)
+        out = values[slc]
+        ctx[op.outputs[-1].key] = out
+
+
+class DataFramePSRSGroupbySample(DataFramePSRSChunkOperand, DataFrameOperandMixin):
+    _op_type_ = OperandDef.GROUPBY_SORT_REGULAR_SAMPLE
+
+    @property
+    def output_limit(self):
+        return 1
+
+    @classmethod
+    def execute(cls, ctx, op):
+        a = ctx[op.inputs[0].key][0]
+        xdf = pd if isinstance(a, (pd.DataFrame, pd.Series)) else cudf
+        if isinstance(a, xdf.Series) and op.output_types[0] == OutputType.dataframe:
+            a = _series_to_df(a, xdf)
+
+        n = op.n_partition
+        if a.shape[0] < n:
+            num = n // a.shape[0] + 1
+            a = xdf.concat([a] * num).sort_index()
+
+        w = a.shape[0] * 1.0 / (n + 1)
+
+        slc = np.linspace(max(w - 1, 0), a.shape[0] - 1, num=n, endpoint=False).astype(
+            int
+        )
+
+        out = a.iloc[slc]
+        if op.output_types[0] == OutputType.series and out.ndim == 2:
+            assert out.shape[1] == 1
+            out = out.iloc[:, 0]
+        ctx[op.outputs[-1].key] = out
+
+
+class DataFrameGroupbySortShuffle(MapReduceOperand, DataFrameOperandMixin):
+    _op_type_ = OperandDef.GROUPBY_SORT_SHUFFLE
+
+    # for shuffle map
+    _by = ListField("by")
```

Comment:
```suggestion
    by = ListField("by")
```

Now we prefer using public property instead of private one, and the `@property` stuffs are not required any more.

Reply:
Fixed @qinxuye. Please confirm.